Add AlertComponent for displaying alerts in the GUI

Rishi-0007 opened the pull request:
This pull request introduces the `Alert.tsx` to the Taipy framework and adds relevant tests. The component allows rendering of customizable alerts with dynamic `message`, `severity`, and `variant` properties. Additionally, there have been significant changes to existing components as per the maintainers' request.

This PR addresses the issue outlined in feature #693.

#### **Changes in this PR**:
- **Added `Alert.tsx`**: A new component that renders an alert using Material-UI's `Alert` component.
- **Renamed Components**: As requested by the maintainers, the existing `Alert.tsx` and `Alert.spec.tsx` have been renamed to `Notification.tsx` and `Notification.spec.tsx` to avoid naming conflicts.
- **Modified `factory.py`**: Registered the `Alert Component` with properties (`message`, `severity`, `variant`) and enabled dynamic updates.
- **Updated `viselements.json`**: Added the definition for the `Alert Component`, including its properties.
- **Added tests**: Created new tests for the `Alert Component` to dynamically verify its behavior for different alert messages, severities, and variants.
- **All Tests Passing**: Ensured that all existing and new tests are passing successfully.

#### **Example test**:
[doc/gui/examples/Alert.py](https://github.com/Avaiga/taipy/pull/1975/files#diff-9804230efd3f6418f556e16878e9b414a33656d12a207e76a223b2fc18bccc6d):
```python
import random
from taipy.gui import Gui

# Initial dynamic properties
severity = "error"
variant = "filled"
message = "This is an error message."
render = True

# A simple page that binds the Alert component to dynamic variables and adds a button to trigger the update
page = """
<|{message}|alert|severity={severity}|variant={variant}|render={render}|>
<br/>
<|Update Alert|button|on_action=update_alert|>
"""

# Function to toggle between variants and severities
def update_alert(state):
    severities = ["error", "warning", "info", "success"]
    variants = ["filled", "outlined"]
    
    state.severity = random.choice(severities)
    state.variant = random.choice(variants)
    state.message = f"This is a {state.severity} message with {state.variant} variant."

if __name__ == "__main__":
    gui = Gui(page)
    gui.run(title="Test Alert")
```
#### **Output**:
https://github.com/user-attachments/assets/ed668b82-410e-42f9-aae1-560b13007104

#### **Additional Information**:
- Followed the coding standards and branch naming conventions.
- All tests have passed, and the component is working as expected.

---

### **Checklist**:
- [x] The code follows the Taipy coding style.
- [x] Tests are included.
- [x] Code is rebased and up to date with the latest `develop` branch.
- [x] Pre-commit hooks are passing without errors.
- [x] All Taipy tests pass successfully.

---

### **Related Issue**:
This PR resolves issue #693.

Rishi-0007 committed:
Add AlertComponent for displaying alerts in the GUI

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

Rishi-0007 commented:
@quest-bot loot #693

Rishi-0007 committed:
Add AlertComponent in index.ts

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/AlertComponent.tsx:
```diff
@@ -0,0 +1,32 @@
+        );
+    }, [severity, message, variant, className, props.id]);
+
+    // Use useMemo to potentially optimize complex rendering, though here it's simple
```
As you wrote, I don't think this is necessary in this simple case

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/AlertComponent.tsx:
```diff
@@ -0,0 +1,32 @@
+    variant?: "filled" | "outlined" | "standard";
+}
+
+const AlertComponent = (props: AlertComponentProps) => {
```
Please renane to Alert

Rishi-0007 replied:
`Alert.tsx` already exists.

FredLL-Avaiga replied:
then we need to rename that to Notification.tsx (and its dependencies)

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/AlertComponent.tsx:
```diff
@@ -0,0 +1,32 @@
+
+interface AlertComponentProps extends TaipyBaseProps {
+    severity?: "error" | "warning" | "info" | "success";
+    message: string;
```
This should be dynamic string 
You can check in others component how dynamic strings are handled

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/AlertComponent.tsx:
```diff
@@ -0,0 +1,32 @@
+    const { severity = "info", message, variant = "filled" } = props;
+
+    // Handle className using a custom hook, if applicable
+    const className = useClassNames(props.libClassName, props.dynamicClassName, props.className);
```
👍

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/index.ts:
```diff
@@ -76,6 +77,7 @@ export const getRegisteredComponents = () => {
             Toggle: Toggle,
             TreeView: TreeView,
             Progress: Progress,
+            AlertComponent: AlertComponent
```
Alphabetical order is easier to read 😎

FredLL-Avaiga commented on taipy/gui/_renderers/factory.py:
```diff
@@ -618,6 +619,19 @@ class _Factory:
+            element_name="AlertComponent",  
+            attributes=attrs,
+        )
+        .set_value_and_default(with_update=True)  # Enable dynamic updates for the 'message'
```
👍I think true is the default value ?

You should give it a type `var_type= PropertyType.dynamic_string`

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/test-alert.py:
This file shouldn't be in the PR.
But you can add simple usage examples in `doc/gui/examples/`

Rishi-0007 committed:
Refactor AlertComponent to handle dynamic message and update dependencies

Rishi-0007 commented:
Hi @FredLL-Avaiga ,

I've implemented the requested changes:
- Removed test from PR.
- Made `message` a dynamic string.
- Removed unnecessary memoization.
- Updated `viselements.json` and `factory.py`.

Please review the changes when you have a moment. Let me know if anything else is needed.

Thanks!

Rishi-0007 committed:
Renamed AlertComponent to Notification and updated dependencies

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

Rishi-0007 commented:
Hi @FredLL-Avaiga ,
Please review the changes when you get a chance. Let me know if any further adjustments are needed.

jrobinAV commented:
@Rishi-0007 If you have a screenshot (or even a few) to illustrate your proposal, that would make the review easier.

In any case, thank you for your help.

Rishi-0007 commented:
> @Rishi-0007 If you have a screenshot (or even a few) to illustrate your proposal, that would make the review easier.
> 
> In any case, thank you for your help.

I have added `Notification.tsx` and its dependencies to solve the issue #693.

But on running this `test.py`
```python
from taipy.gui import Gui, State

# Initialize state variables for the notification component
notification_message = "This is a notification message!"
notification_severity = "info"
notification_variant = "filled"

# Function to handle changes to the message, severity, or variant
def on_change(state: State, var_name: str, value: str):
    if var_name == "notification_message":
        state.notification_message = value
    elif var_name == "notification_severity":
        state.notification_severity = value
    elif var_name == "notification_variant":
        state.notification_variant = value
    print(f"Updated {var_name} to: {value}")

# Markdown definition to render the Notification component and input controls
md = """
# Notification Component Test

## Current Notification:
<|{notification_message}|notification|severity={notification_severity}|variant={notification_variant}|>

### Update Notification Message:
<|{notification_message}|input|on_change=on_change|label=Update Message|class_name=fullwidth|>

### Update Notification Severity:
<|{notification_severity}|selector|lov=error;warning;info;success|on_change=on_change|label=Update Severity|class_name=row fullwidth|>

### Update Notification Variant:
<|{notification_variant}|selector|lov=filled;outlined;standard|on_change=on_change|label=Update Variant|class_name=row fullwidth|>
"""

# Run the GUI
Gui(md).run()

```
i am getting this error.

#### Screenshot:
![image](https://github.com/user-attachments/assets/b2ac1f69-e763-4511-abf0-e972073f6082)

FredLL-Avaiga commented on taipy/gui/_renderers/factory.py:
```diff
@@ -618,6 +619,19 @@ class _Factory:
+        .set_attributes(
+            [
+                ("severity", PropertyType.string),  # severity is a simple string property
+                ("message", PropertyType.dynamic_string),  # message is now a dynamic string
```
message shouldn't be described here
you need to pass var_type= PropertyType.dynamic_string to set_value_and_default() line 628

FredLL-Avaiga commented on taipy/gui/_renderers/factory.py:
```diff
@@ -618,6 +619,19 @@ class _Factory:
             ]
         )
         ._set_propagate(),
+        "notification": lambda gui, control_type, attrs: _Builder(
```
this can and should stay "alert"

FredLL-Avaiga commented on taipy/gui/viselements.json:
```diff
@@ -1735,7 +1735,36 @@
-        ]
+        ],
+        [
+            "notification",
```
"alert" also here

Rishi-0007 committed:
updated dependencies

Rishi-0007 commented:
Hi @FredLL-Avaiga,
Please review the latest changes and let me know if any further adjustments are needed.

FredLL-Avaiga commented on taipy/gui/viselements.json:
```diff
@@ -1746,25 +1746,18 @@
-                        "doc": "Defines the severity level of the alert. Valid values: \"error\", \"warning\", \"info\", \"success\"."
-                    },
-                    {
-                        "name": "message",
```
message property still needs to be described

Rishi-0007 committed:
updated viselements.json

Rishi-0007 commented:
Hey @FredLL-Avaiga ,

I've updated the message property in `viselements.json` as requested, describing it as a dynamic string. Apologies for the earlier oversight. Please review the changes when you have a moment.

namnguyen20999 commented on taipy/gui/viselements.json:
```diff
@@ -1735,7 +1735,36 @@
+                "properties": [
+                    {
+                        "name": "severity",
+                        "default_property": false,
```
It is not necessary to define "default_property" here if it is not default.

namnguyen20999 commented on taipy/gui/_renderers/factory.py:
```diff
@@ -618,6 +619,18 @@ class _Factory:
+        .set_attributes(
+            [
+                ("severity", PropertyType.string),  # severity is a simple string property
+            ]
```
I guess you still missing the "variant" property here

namnguyen20999 commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
```diff
@@ -0,0 +1,24 @@
+
+interface NotificationProps extends TaipyBaseProps {
+    severity?: "error" | "warning" | "info" | "success";
+    message: string | (() => string);  // Dynamic string handling
```
add defaultMessage?: string here

namnguyen20999 commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
```diff
@@ -0,0 +1,24 @@
+    const { severity = "info", message, variant = "filled" } = props;
+
+    // Handle dynamic string by checking if `message` is a function and calling it
+    const displayMessage = typeof message === "function" ? message() : message;
```
const displayMessage = useDynamicProperty(props.message, props.defaultMessage, undefined); you could take a look into our other components to see how dynamic properties are being handled

namnguyen20999 reviewed (changes requested):
almost there

Rishi-0007 committed:
Refactor Notification component to handle dynamic message and add defaultMessage property

FredLL-Avaiga commented:
Hi @Rishi-0007, can you fix the conflicts and resolve the conversation that are no longer active ?

Rishi-0007 commented:
Hi @FredLL-Avaiga,

Thank you for your feedback. I'm not entirely sure about the best way to resolve the conflicts and manage the conversations that are no longer active. Could you kindly provide some guidance on how to proceed with these tasks?

I appreciate your help!

FredLL-Avaiga commented:
> Hi @FredLL-Avaiga,
> 
> Thank you for your feedback. I'm not entirely sure about the best way to resolve the conflicts and manage the conversations that are no longer active. Could you kindly provide some guidance on how to proceed with these tasks?
> 
> I appreciate your help!

I'm sure Git and GitHub docs are very good 😊

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

Rishi-0007 commented:
Hi @FredLL-Avaiga @namnguyen20999 ,

I've followed the steps for adding the `Notification` component, but I'm still encountering an issue where the component is not recognized. 

#### `test`
```python
from taipy.gui import Gui

# Define the dynamic properties for the notification
alert_message = "This is a dynamic notification!"
alert_severity = "info"
alert_variant = "filled"

# Define the page with the notification component
page = """
<|{alert_message}|alert|severity={alert_severity}|variant={alert_variant}|>
"""

if __name__ == "__main__":
    Gui(page).run(title="Notification Test")
```
#### output:

`Unrecognized control alert at line 2: "<|{alert_message}|alert|severity={alert_severity}|variant={alert_variant}|>"`.

![image](https://github.com/user-attachments/assets/4b11dc43-5fc4-45b8-b77a-6252016a8630)

Despite trying multiple fixes, the issue persists. Could you please help me troubleshoot this?

Thank you!

---

Let me know if you'd like to adjust anything further!

FabienLelaquais commented on frontend/taipy-gui/src/components/Taipy/index.ts:
```diff
@@ -43,6 +43,7 @@ import Table from "./Table";
 import TaipyStyle from "./TaipyStyle";
 import Toggle from "./Toggle";
 import TreeView from "./TreeView";
+import Notification from "./Notification";
```
In the nearest possible future, we will have to rename 'Notification' to 'Alert' for homogeneity (although programmers don't really care).

Rishi-0007 replied:
Hi @FabienLelaquais thanks for your review, but `Alert.tsx` already exists.

FabienLelaquais commented on taipy/gui/viselements.json:
```diff
@@ -1735,7 +1735,40 @@
+                        "name": "severity",
+                        "type": "dynamic(str)",
+                        "default_value": "\"error\"",
+                        "doc": "Defines the severity level of the notification. Valid values: \"error\", \"warning\", \"info\", \"success\"."
```
```suggestion
                        "doc": "The severity level of the alert. Valid values: \"error\", \"warning\", \"info\", \"success\".\nThe default is \"error\"."
```

FabienLelaquais commented on taipy/gui/viselements.json:
```diff
@@ -1735,7 +1735,40 @@
+                        "doc": "The message displayed in the notification. Can be a dynamic string."
+                    },
+                    {
+                        "name": "variant",
```
I'm not sure we need this 'variant' property, although it simplifies styling.
I mean no other element has that kind of styling=-dedicated property.
We will discuss that.

FabienLelaquais commented on taipy/gui/viselements.json:
```diff
@@ -1735,7 +1735,40 @@
+                        "name": "variant",
+                        "type": "str",
+                        "default_value": "\"filled\"",
+                        "doc": "Defines the variant of the notification. Valid values: \"filled\", \"outlined\", \"standard\"."
```
```suggestion
                        "doc": "The variant of the alert. Valid values: \"filled\", \"outlined\", \"standard\".\nThe default is \"filled\"."
```

FabienLelaquais commented on taipy/gui/viselements.json:
```diff
@@ -1735,7 +1735,40 @@
+                        "doc": "Defines the variant of the notification. Valid values: \"filled\", \"outlined\", \"standard\"."
+                    },
+                    {
+                        "name": "defaultMessage",
```
You don't want this.
'defaultMessage' was not meant to be an element property, but a property for the React component: it is set to the initial value of the 'message' React property, when no binding has yet initialized for for display.

FabienLelaquais commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
To be renamed to 'Alert.tsx' when we're done with the code itself.

Rishi-0007 replied:
Done, please check it.

FabienLelaquais commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
```diff
@@ -0,0 +1,31 @@
+    severity?: "error" | "warning" | "info" | "success";
+    message: string | (() => string);  // Dynamic string handling
+    variant?: "filled" | "outlined" | "standard";
+    defaultMessage?: string;
```
Here is exactly where 'defaultMessage' would appear, and not on the back-end side.
That is relevant.

FredLL-Avaiga replied:
a example of using dynamic property: 
```ts
interface CompProps {
  value?: string;
  defaultValue?: string;
}
```
In the body of you component:
```ts
const value = useDynamicProperty(props.value, props.defaultValue, "");
```

on the backend side
if value is default property
```py
.set_value_and_default() (var_type=dynamic_string is default)
```
else
```py
.set_attributes(
            [
                ("value", PropertyType.dynamic_string),
...
```

FabienLelaquais commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
```diff
@@ -0,0 +1,31 @@
+    defaultMessage?: string;
+}
+
+const Notification = (props: NotificationProps) => {
```
To be renamed to 'Alert' whenever.

FabienLelaquais commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
```diff
@@ -0,0 +1,31 @@
+    let displayMessage = useDynamicProperty(props.message, props.defaultMessage, undefined);
+
+    // Ensure displayMessage is a string
+    if (typeof displayMessage === "function") {
```
I don't understand what your goal is here.
You're not testing if it's a string. You're testing if it's a function...

FredLL-Avaiga replied:
no need for this, it's all taken care of by `useDynamicProperty`

FabienLelaquais commented on taipy/gui/_renderers/factory.py:
```diff
@@ -618,6 +619,20 @@ class _Factory:
+            [
+                ("severity", PropertyType.string),  
+                ("variant", PropertyType.string),  
+                ("defaultMessage", PropertyType.string),
```
Remove this

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
```diff
@@ -0,0 +1,31 @@
+
+interface NotificationProps extends TaipyBaseProps {
+    severity?: "error" | "warning" | "info" | "success";
+    message: string | (() => string);  // Dynamic string handling
```
should be declared as `message?: string;`

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
```diff
@@ -0,0 +1,31 @@
+    const { severity = "info", variant = "filled" } = props;
+
+    // Use useDynamicProperty to handle dynamic message and defaultMessage
+    let displayMessage = useDynamicProperty(props.message, props.defaultMessage, undefined);
```
use const whenever possible
Third argument should be `""` as it is passed as a children of the Alert Mui component

FredLL-Avaiga reviewed (changes requested):
UNderstanding dynamic property is not easy

Rishi-0007 committed:
Apply suggestions from code review

Co-authored-by: Fabien Lelaquais <86590727+FabienLelaquais@users.noreply.github.com>

Rishi-0007 committed:
Refactor Notification component and add dynamic message handling

- Refactor the Notification component in Taipy to improve code readability and maintainability.
- Add support for dynamic message handling using the useDynamicProperty hook.
- Remove the variant and defaultMessage properties from the Notification component, as they are no longer needed.

Closes #693

Rishi-0007 commented:
Hi @FredLL-Avaiga, @namnguyen20999, and @FabienLelaquais,

I've made the requested changes, and I sincerely apologize for the earlier mistakes—this is my first time contributing to open source. Thanks to your feedback, I figured out the issue: I had installed Taipy via `pip` instead of locally.

I've learned a lot from this process! Here's a test example to check the `Notification` component:

```python
from taipy.gui import Gui

alert_message = "This is an alert message!"
alert_severity = "error"

page = """
<|{alert_message}|alert|severity={alert_severity}|>
"""

Gui(page).run(title="Notification Test")
```

Screenshot:
![image](https://github.com/user-attachments/assets/cdc0a7df-b917-4e04-a65a-c90ab2c11739)

---

Let me know if further changes are needed.
Thank you!

FredLL-Avaiga commented:
> Hi @FredLL-Avaiga, @namnguyen20999, and @FabienLelaquais,
> 
> I've made the requested changes, and I sincerely apologize for the earlier mistakes—this is my first time contributing to open source. Thanks to your feedback, I figured out the issue: I had installed Taipy via `pip` instead of locally.
> 
> I've learned a lot from this process! Here's a test example to check the `Notification` component:
> 
> ```python
> from taipy.gui import Gui
> 
> alert_message = "This is an alert message!"
> alert_severity = "error"
> 
> page = """
> <|{alert_message}|alert|severity={alert_severity}|>
> """
> 
> Gui(page).run(title="Notification Test")
> ```
> 
> Screenshot:
> ![image](https://github.com/user-attachments/assets/cdc0a7df-b917-4e04-a65a-c90ab2c11739)
> 
> ---
> 
> Let me know if further changes are needed.
> Thank you!

Every mistake is good if it helps you learn something!

FredLL-Avaiga commented on frontend/taipy/package.json:
Please do not commit this change

Rishi-0007 replied:
Okay i got it thanks for reviewing :+1:

FredLL-Avaiga replied:
no commit for this file: it should *not* be changed

FredLL-Avaiga reviewed (changes requested):
Looks very good to me
Last step would be to rename Alert.tsx to Notification.tsx and vice versa
Tx for your work

FabienLelaquais commented on frontend/taipy/package.json:
```diff
@@ -30,7 +30,8 @@
     "react": "^18.2.0",
-    "react-dom": "^18.2.0"
+    "react-dom": "^18.2.0",
+    "taipy-gui": "file:../../taipy/gui/webapp"
```
Do not commit that either

FabienLelaquais commented on taipy/gui/viselements.json:
```diff
@@ -1,5 +1,26 @@
 {
     "controls": [
+        [
+            "alert",  
```
So you moved the definition of the control to the top.
Actually we use the order of the declaration to generate the documentation.
We want 'more used' elements to appear first.
To me, this should go just after "status".

And same for the properties: more used come first.
In your case the order should be "message", "severity".

FabienLelaquais replied:
And where is variant gone?
I guess we can do that with styling, and that's fine.

FabienLelaquais commented on taipy/gui/viselements.json:
```diff
@@ -1,5 +1,26 @@
+                "properties": [
+                    {
+                        "name": "severity",
+                        "type": "dynamic(str)",
```
Nope. Not dynamic - according to the code.

FabienLelaquais commented:
@Rishi-0007 I read your apologies. You really don't have to.
This part is one of the tricky one in Taipy GUI, so apologies accepted, but useless.

@FredLL-Avaiga I wonder if the severity should not be dynamic after all.
A page would typically hold a single of these controls, applicable to different conditions.
But then... how do we mask it?
Shouldn't we have a 'render' property like in parts?

Rishi-0007 committed:
Renamed Notification component to Alert and vice versa and updated dependencies

Rishi-0007 committed:
Refactor Alert component and update dependencies

Rishi-0007 commented:
Hi @FredLL-Avaiga @FabienLelaquais,

I’ve successfully committed the requested changes and resolved the conflicts. 

#### here is the final output:
![image](https://github.com/user-attachments/assets/f2d53edd-c26f-439a-9507-8cee72fce1d3)

---
Let me know if further changes are needed.

Rishi-0007 reviewed (commented):
Made suggested changes.

FredLL-Avaiga commented:
>  I wonder if the severity should not be dynamic after all.
> A page would typically hold a single of these controls, applicable to different conditions.
> But then... how do we mask it?
> Shouldn't we have a 'render' property like in parts?

I think you're right, good thinking!
Then all properties should be dynamic and we need to add a render dynamic boolean property.

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

Rishi-0007 commented:
Hi @FredLL-Avaiga, @namnguyen20999, @FabienLelaquais,

Thank you for your feedback and insights! Based on your discussion, I agree that making the `severity` property dynamic and adding a `render` dynamic boolean property would enhance flexibility. This way, the component can adapt to different conditions and be shown/hidden like parts. If you're aligned with this approach, I’d be happy to implement the changes and make all the properties dynamic as well.

Let me know your thoughts!

FredLL-Avaiga commented:
> Hi @FredLL-Avaiga, @namnguyen20999, @FabienLelaquais,
> 
> Thank you for your feedback and insights! Based on your discussion, I agree that making the `severity` property dynamic and adding a `render` dynamic boolean property would enhance flexibility. This way, the component can adapt to different conditions and be shown/hidden like parts. If you're aligned with this approach, I’d be happy to implement the changes and make all the properties dynamic as well.
> 
> Let me know your thoughts!

Yes go for it

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

Rishi-0007 committed:
Refactor Alert component to add dynamic rendering capability

Rishi-0007 commented:
hey @FredLL-Avaiga @FabienLelaquais ,
I have made the suggested changes. Please check it and let me know if any further changes are needed.
Thank you!

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
```diff
@@ -0,0 +1,84 @@
+    alerts: AlertMessage[];
+}
+
+const Alert = ({ alerts }: AlertProps) => {
```
this is `Notification`

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Alert.tsx:
```diff
@@ -1,84 +1,29 @@
-    const lastKey = useRef<SnackbarKey>("");
-    const { enqueueSnackbar, closeSnackbar } = useSnackbar();
-    const dispatch = useDispatch();
+const Notification = (props: NotificationProps) => {
```
this is `Alert`

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/index.ts:
```diff
@@ -13,6 +13,7 @@
 
 import { ComponentType } from "react";
 
+import Notification from "./Alert";
```
then this would be `import Alert from "./Alert";`

FredLL-Avaiga commented on taipy/gui/_renderers/factory.py:
```diff
@@ -69,6 +70,21 @@ class _Factory:
+        .set_value_and_default(var_type=PropertyType.dynamic_string) 
+        .set_attributes(
+            [
+                ("severity", PropertyType.string),
```
dynamic string

FredLL-Avaiga commented on taipy/gui/_renderers/factory.py:
```diff
@@ -69,6 +70,21 @@ class _Factory:
+        .set_attributes(
+            [
+                ("severity", PropertyType.string),
+                ("variant", PropertyType.string),
```
dynamic string

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Alert.tsx:
```diff
@@ -1,84 +1,29 @@
+import { TaipyBaseProps } from "./utils";
+import { useDynamicProperty } from "../../utils/hooks";
+
+interface NotificationProps extends TaipyBaseProps {
```
AlertProps
missing props:
- defaultSeverity?: string;
- defaultVariant?: string;
- defaultRender?: boolean;

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Alert.tsx:
```diff
@@ -1,84 +1,29 @@
-    const { enqueueSnackbar, closeSnackbar } = useSnackbar();
-    const dispatch = useDispatch();
+const Notification = (props: NotificationProps) => {
+    const render = useDynamicProperty(props.render, true, true);
+    const severity = useDynamicProperty(props.severity, "error", "error");
+    const variant = useDynamicProperty(props.variant, "filled", "filled");
+    const message = useDynamicProperty(props.message, props.defaultMessage, "");
```
```
const render = useDynamicProperty(props.render, props.defaultRender, true);
const severity = useDynamicProperty(props.severity, props.defaultSeverity, "error");
const variant = useDynamicProperty(props.variant, props.defaultVariant, "filled");
const message = useDynamicProperty(props.message, props.defaultMessage, "");
```

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Alert.tsx:
```diff
@@ -1,84 +1,29 @@
-        ),
-        [resetAlert]
+    return (
+        <Alert severity={severity} variant={variant} id={props.id} className={props.className}>
```
className needs to be dynamic too
const className = useClassNames(props.libClassName, props.dynamicClassName, props.className);

Rishi-0007 committed:
feat: Enhance TaipyAlert with dynamic classNames and dispatch actions

- Added dynamic className handling to TaipyAlert component.
- Implemented dispatching of update actions.
- Fixed issues with severity and variant properties.
- Added unit tests to validate the new behavior, with all tests passing successfully.

Rishi-0007 commented:
Hi @FredLL-Avaiga,

I have made the requested changes based on your review. Please take a look at the updated output. I have added `Alert.spec.tsx` the tests, and I'm happy to report that all tests passed successfully. Also, I apologize for the earlier mistakes and appreciate your feedback.

`test.py`:
```python
import random
from taipy.gui import Gui

# Initial dynamic properties
severity = "error"
variant = "filled"
message = "This is an error message."
render = True

# A simple page that binds the Alert component to dynamic variables and adds a button to trigger the update
page = """
<|{message}|alert|severity={severity}|variant={variant}|render={render}|>
<br/>
<|Update Alert|button|on_action=update_alert|>
"""

# Function to toggle between variants and severities
def update_alert(state):
    severities = ["error", "warning", "info", "success"]
    variants = ["filled", "outlined"]
    
    state.severity = random.choice(severities)
    state.variant = random.choice(variants)
    state.message = f"This is a {state.severity} message with {state.variant} variant."

if __name__ == "__main__":
    gui = Gui(page)
    gui.run(title="Test Alert")
```
#### Output:

https://github.com/user-attachments/assets/016a7181-f755-4087-a7a6-ec108cc28745

_browser needs to be refresh to see the changes_

FredLL-Avaiga commented:
`browser needs to be refresh to see the changes`
That's really bad and not what were want
Or did i misunderstood something?

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

Rishi-0007 commented:
> `browser needs to be refresh to see the changes` That's really bad and not what were want Or did i misunderstood something?

Hi @FredLL-Avaiga ,
thanks for pointing out this i thought it was issue of my test.py but after investigating, I noticed that changes to the alert's properties only reflect after a browser refresh, and state updates don't seem to trigger a re-render as expected. Could this be an issue with the useDynamicProperty hook or state management?

I'd appreciate your insights or any suggestions for resolving this.

FredLL-Avaiga commented:
I'll have to dig a bit later

On Mon, Oct 14, 2024, 21:01 Rishi Nayak ***@***.***> wrote:

> browser needs to be refresh to see the changes That's really bad and not
> what were want Or did i misunderstood something?
>
> Hi @FredLL-Avaiga <https://github.com/FredLL-Avaiga> ,
> thanks for pointing out this i thought it was issue of my test.py but
> after investigating, I noticed that changes to the alert's properties only
> reflect after a browser refresh, and state updates don't seem to trigger a
> re-render as expected. Could this be an issue with the useDynamicProperty
> hook or state management?
>
> I'd appreciate your insights or any suggestions for resolving this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Avaiga/taipy/pull/1975#issuecomment-2412014954>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVQBA5A3Y46BFTTHIQ4RH33Z3QIIBAVCNFSM6AAAAABPTB2TX6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMJSGAYTIOJVGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

FredLL-Avaiga commented:
> ```python
> import random
> from taipy.gui import Gui
> 
> # Initial dynamic properties
> severity = "error"
> variant = "filled"
> message = "This is an error message."
> render = True
> 
> # A simple page that binds the Alert component to dynamic variables and adds a button to trigger the update
> page = """
> <|{message}|alert|severity={severity}|variant={variant}|render={render}|>
> <br/>
> <|Update Alert|button|on_action=update_alert|>
> """
> 
> # Function to toggle between variants and severities
> def update_alert(state):
>     severities = ["error", "warning", "info", "success"]
>     variants = ["filled", "outlined"]
>     
>     state.severity = random.choice(severities)
>     state.variant = random.choice(variants)
>     state.message = f"This is a {state.severity} message with {state.variant} variant."
> 
> if __name__ == "__main__":
>     gui = Gui(page)
>     gui.run(title="Test Alert")
> ```

Works for me

https://github.com/user-attachments/assets/70b679fa-771c-4cd9-969f-48eb3c740c27

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

Rishi-0007 commented:
Hi @FredLL-Avaiga , 
I have also tested the `Alert` component in the browser, and I can confirm that all the properties (severity, variant, message, etc.) are working as intended without needing a refresh. I guess the issue is resolved here. Could you please let me know the next steps, or if there’s anything else that needs to be addressed?

https://github.com/user-attachments/assets/ed668b82-410e-42f9-aae1-560b13007104


Thanks for your feedback and support!

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
```diff
@@ -0,0 +1,84 @@
+    alerts: AlertMessage[];
+}
+
+const TaipyNotification = ({ alerts }: NotificationProps) => {
```
Why not call it `Notification` ?

FredLL-Avaiga replied:
?

FredLL-Avaiga commented on frontend/taipy/package.json:
please do not commit this file

FredLL-Avaiga replied:
I'd like you to *NOT* commit this file

FredLL-Avaiga reviewed (changes requested):
Looks very good to me
Can you add one or two very simple examples in doc/gui/examples ?

Rishi-0007 committed:
Add Alert.py example with dynamic properties and button to update alert

FredLL-Avaiga commented on doc/gui/examples/Alert.py:
```diff
@@ -0,0 +1,43 @@
+severity = "error"
+variant = "filled"
+message = "This is an error message."
+render = True
```
I wouldn't use render here

FredLL-Avaiga commented on doc/gui/examples/Alert.py:
```diff
@@ -0,0 +1,43 @@
+# A simple page that binds the Alert component to dynamic variables and adds a button to trigger the update
+page = """
+<|{message}|alert|severity={severity}|variant={variant}|render={render}|>
+<br/>
```
no need for br or button

FredLL-Avaiga commented on doc/gui/examples/Alert.py:
```diff
@@ -0,0 +1,43 @@
+<|Update Alert|button|on_action=update_alert|>
+"""
+
+# Function to toggle between variants and severities
```
no need for a button on_action

FredLL-Avaiga commented on doc/gui/examples/Alert.py:
```diff
@@ -0,0 +1,43 @@
+    state.message = f"This is a {state.severity} message with {state.variant} variant."
+
+if __name__ == "__main__":
+    gui = Gui(page)
```
this can fit in one line
don't forget to add a blank line at the end of the file

Rishi-0007 commented:
@FredLL-Avaiga  does this looks good to you?
```python
from taipy.gui import Gui

severity = "error"
variant = "filled"
message = "This is an error message."

page = """
<|{message}|alert|severity={severity}|variant={variant}|>
"""

if __name__ == "__main__":
    gui = Gui(page)
    gui.run(title="Test Alert")
```

FredLL-Avaiga commented:
You could even be not simplistic and keep only the message as a variable and group the 2 lines with gui

Rishi-0007 commented:
> You could even be not simplistic and keep only the message as a variable and group the 2 lines with gui

```python
from taipy.gui import Gui

message = "This is an error message."

Gui(f"<|{{message}}|alert|severity='error'|variant='filled'|>").run(title="Test Alert")

```

Is this gd?

FredLL-Avaiga commented:
> > You could even be not simplistic and keep only the message as a variable and group the 2 lines with gui
> 
> ```python
> from taipy.gui import Gui
> 
> message = "This is an error message."
> 
> Gui(f"<|{{message}}|alert|severity='error'|variant='filled'|>").run(title="Test Alert")
> 
> ```
> 
> Is this gd?

A bit too compact

Rishi-0007 committed:
Refactor Alert.py example and add package.json

- Refactor Alert.py example to remove unused code and simplify the page structure.
- Add package.json file for frontend/taipy directory from develop branch.

Rishi-0007 committed:
refactor package.json to match it with develop branch

Rishi-0007 commented:
Hey @FredLL-Avaiga , 
I've added a simple test :
```python
from taipy.gui import Gui

severity = "error"
variant = "filled"
message = "This is an error message."

page = """
<|{message}|alert|severity={severity}|variant={variant}|>
"""

if __name__ == "__main__":
    gui = Gui(page)
    gui.run(title="Test Alert")

```
output:
![Screenshot from 2024-10-20 02-36-32](https://github.com/user-attachments/assets/66b2765f-7320-4bab-bdab-55a85fa470fe)


Also i have removed package.json from my commit.
Thank you for your time. I hope everything is resolved now.

FredLL-Avaiga reviewed (changes requested):
I'd like you to *NOT* commit any package.json file

FredLL-Avaiga reviewed (changes requested):
Almost there

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

FredLL-Avaiga commented:
Waiting for this [PR](https://github.com/Avaiga/taipy/pull/1985) to be merged because it'll be easier to integrate then

Rishi-0007 commented:
> Please avoid committing the `package.json` file.

![Screenshot from 2024-10-21 22-12-36](https://github.com/user-attachments/assets/996bb11d-b5e1-4fa6-a927-52218f13ef14)

I’ve checked, and the `package.json` file no longer appears. Kindly review, and let me know if any additional changes are needed to proceed with the merge.

Rishi-0007 commented:
> Waiting for this [PR](https://github.com/Avaiga/taipy/pull/1985) to be merged because it'll be easier to integrate then

Thanks for the update! I understand, and I’ll wait for the PR to be merged.

FredLL-Avaiga committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

Rishi-0007 committed:
Add license headers to Alert components

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

Rishi-0007 committed:
Fixed linter issue using ruff

FredLL-Avaiga commented:
The PR about closing notification (that has an impact on the old alert component) was merged.
I'm sorry but can you manage the merge and conflicts resolution ?
tx @Rishi-0007

FredLL-Avaiga reviewed (changes requested):
all good except for the conflicts with the last PR merged

Rishi-0007 committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
```diff
@@ -0,0 +1,84 @@
+        if (alert) {
+            if (alert.atype === "") {
+                if (lastKey.current) {
+                    closeSnackbar(lastKey.current);
```
Something was lost in the conflict resolution
this should be
```ts
    useEffect(() => {
        if (alert) {
             const notificationId = alert.notificationId || "";
            if (alert.atype === "") {
                closeSnackbar(notificationId);  
            } else {
                enqueueSnackbar(alert.message, {
                    variant: alert.atype as VariantType,
                    action: notifAction,  
                    autoHideDuration: alert.duration,
                    key: notificationId,  
                });
                alert.system && new Notification(document.title || "Taipy", { body: alert.message, icon: faviconUrl });
            }
            dispatch(createDeleteAlertAction(notificationId));
        }
    }, [alert, enqueueSnackbar, closeSnackbar, notifAction, faviconUrl, dispatch]);

```

Rishi-0007 committed:
Refactor Notification component and fix merge issue

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Notification.tsx:
```diff
@@ -0,0 +1,82 @@
+
+    useEffect(() => {
+        if (alert) {
+            const notificationId = nanoid();
```
There was a miss on the other PR and this should be:
`const notificationId = alert.notificationId || "";`

Rishi-0007 committed:
Refactor Notification component and fix issue due to other PR

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Notification.spec.tsx:
```diff
@@ -0,0 +1,202 @@
+    });
+
+    it("Alert disappears when alert type is empty", async () => {
+        const alerts = [{ atype: "success", message: "Test Alert", duration: 3000, system: false }];
```
should be `const alerts = [{ atype: "success", message: "Test Alert", duration: 3000, system: false, notificationId: "aNotificationId" }];`

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/Notification.spec.tsx:
```diff
@@ -0,0 +1,202 @@
+            </SnackbarProvider>,
+        );
+        await screen.findByRole("button", { name: /close/i });
+        const newAlerts = [{ atype: "", message: "Test Alert", duration: 3000, system: false }];
```
and should be  `const newAlerts = [{ atype: "", message: "Test Alert", duration: 3000, system: false, notificationId: "aNotificationId" }];`

FredLL-Avaiga reviewed (changes requested):
another miss in the previous PR :-(

FredLL-Avaiga committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

Rishi-0007 committed:
Refactor Notification test component to include notificationId in alerts

Rishi-0007 committed:
Merge branch 'feature/#693-AddAlertVisualElement' of https://github.com/Rishi-0007/taipy into feature/#693-AddAlertVisualElement

FredLL-Avaiga reviewed (approved):
Well done @Rishi-0007

FredLL-Avaiga commented:
@jrobinAV can you merge this ?

Rishi-0007 commented:
> Well done @Rishi-0007

I sincerely appreciate your guidance and support throughout this journey. Contributing to this organization has been a tremendous learning experience for me. Thank you so much for everything!

FredLL-Avaiga committed:
Merge branch 'develop' into feature/#693-AddAlertVisualElement

RymMichaut commented:
Congratulations, @Rishi-0007, for this merged Hacktoberfest PR.
Can you please contact me at [rym.michaut@taipy.io](mailto:rym.michaut@taipy.io) to prepare your swag and send it to you ASAP.
